apps/src/blockly/addons/cdoSerializationHelpers.js

```diff
@@ -0,0 +1,110 @@
+import BlockSvgUnused from './blockSvgUnused';
+
+const SORT_BY_POSITION = true;
+
+/**
+ * Converts XML serialization to JSON using a temporary unrendered workspace.
+ * @param {xml} xml - workspace serialization, current/legacy format
+ * @return {json} stateToLoad - modern workspace serialization
+ */
+export function convertXmlToJson(xml) {
+  const tempWorkspace = new Blockly.Workspace();
+
+  // domToBlockSpace returns an array of "block" objects with the following property:
+  //   blockly_block: the actual block object created by Blockly
+  //   x: the x attribute found in <block/> element
+  //   y: the y attribute found in <block/> element
+  const xmlBlocks = Blockly.Xml.domToBlockSpace(tempWorkspace, xml);
+
+  const stateToLoad = Blockly.serialization.workspaces.save(tempWorkspace);
+
+  // Create a map of ids (key) and blocks (value).
+  const blockIdMap = new Map();
+  stateToLoad.blocks.blocks.forEach(block => {
+    blockIdMap.set(block.id, block);
+  });
+
+  addPositionsToState(xmlBlocks, blockIdMap);
+  tempWorkspace.dispose();
+  return stateToLoad;
+}
+
+/**
+ * Adds x/y values from XML to JSON serialization.
+ * @param {Array<Object>} xmlBlocks - an array of "block" objects containing a block and x/y coordinates
+ * @param {Map<String, Object>} blockIdMap - a map of ids (keys) and blocks (values)
+ */
+function addPositionsToState(xmlBlocks, blockIdMap) {
+  xmlBlocks.forEach(xmlBlock => {
+    const block = blockIdMap.get(xmlBlock.blockly_block.id);
+    if (block) {
+      // Do not change block values if xmlBlock values are NaN (unspecified in XML)
+      block.x = xmlBlock.x || block.x;
+      block.y = xmlBlock.y || block.y;
+    }
+  });
+}
+
+/**
+ * Use a cursor to position blocks on a workspace (if they do not already have positions)
+ * @param {Blockly.Workspace} workspace - the current Blockly workspace
+ * @param {Map<String, Object>} blockIdMap - a map of ids (keys) and blocks (values)
+ */
+export function positionBlocks(workspace) {
+  const blocks = workspace.getTopBlocks(SORT_BY_POSITION);
+  if (!workspace.rendered) {
+    return blocks;
+  }
+  const metrics = workspace.getMetrics();
+  const contentWidth = metrics.contentWidth;
+  const viewWidth = metrics ? metrics.viewWidth : 0;
+  const padding = viewWidth ? 16 : 0;
+  const verticalSpaceBetweenBlocks = 10;
+
+  // The "cursor" object tracks a position on the workspace that starts in the top
+  // corner, then moves below each block that is manually repositioned.
+  let cursor = {
+    x: padding,
+    y: padding,
+  };
+  if (workspace.RTL && viewWidth) {
+    // Position the cursor from the right of the workspace.
+    cursor.x = viewWidth - padding;
+  } else if (workspace.RTL && !viewWidth) {
+    // Position the cursor from the right of the block.
+    cursor.x = contentWidth - padding;
+  }
```

Comment:
Again, non-blocking suggestion, but this could be crunched down into:
`
if (workspace.RTL) {
  cursor.x = (viewWidth || contentWidth) - padding
}
`
I'm not super stoked about it since I'm not crazy about the embedded `||` in the subtraction. Maybe put that in a separate variable?
`
if (workspace.RTL) {
  const width = viewWidth || contentWidth
  cursor.x = width - padding
}
`

Again, there's nothing wrong with the way it currently is, and I can see arguments against the more compact version. I'm just trying to get rid of the slightly redundant logic, though this is a case of diminishing returns on it, I feel.

If `viewWidth` could use `contentWidth` as its default it'd clean it up even more, but I don't think that's the case.

Also - is there any way to type multi-line code blocks in a github comment? :-(

Reply:
@thomasoniii you can do multi-line code blocks with triple ticks (like ```) on their own lines preceding/following the lines containing your code:

```
if (foo) {
  bar()
}
```

Reply:
Could also do a ternary like:
```
const width = viewWidth || contentWidth;
cursor.x = workspace.RTL ? width - padding : cursor.x;

```
Is that more readable @thomasoniii ?

Reply:
Ooh, how's this?

```
  const metrics = workspace.getMetrics() || {contentWidth: 0, viewWidth: 0};
  const contentWidth = metrics.contentWidth;
  const viewWidth = metrics.viewWidth;
  const padding = viewWidth ? 16 : 0;
  const width = viewWidth || contentWidth;

  // The "cursor" object tracks a position on the workspace that starts in the top
  // corner, then moves below each block that is manually repositioned.
  let cursor = {
    x: padding,
    y: padding,
  };
  cursor.x = workspace.RTL ? width - cursor.x : cursor.x;
```

Reply:
I like this latest version - simple and clear.

Reply:
I like it! With one further suggestion to use destructuring to avoid the interim variables:

```
const { contentWidth, viewWidth} = workspace.getMetrics() || { contentWidth : 0, viewWidth : 0};
```

Reply:
Very cool, thanks for the tips!

Reply:
Very cool, thanks for the tips!